receivers/mqtt/mqtt.go

```diff
@@ -0,0 +1,135 @@
+package mqtt
+
+import (
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	mqttLib "github.com/eclipse/paho.mqtt.golang"
+	"github.com/prometheus/alertmanager/notify"
+	"github.com/prometheus/alertmanager/types"
+
+	"github.com/grafana/alerting/logging"
+	"github.com/grafana/alerting/receivers"
+	"github.com/grafana/alerting/templates"
+)
+
+type Client interface {
+	Connect() mqttLib.Token
+	Publish(topic string, qos byte, retained bool, payload interface{}) mqttLib.Token
+	Disconnect(quiesce uint)
+}
+
+type Notifier struct {
+	*receivers.Base
+	log      logging.Logger
+	tmpl     *templates.Template
+	settings Config
+	client   Client
+}
+
+func defaultClientFactory(opts *mqttLib.ClientOptions) Client {
+	return mqttLib.NewClient(opts)
+}
+
+func New(cfg Config, meta receivers.Metadata, template *templates.Template, logger logging.Logger, clientFactory func(opts *mqttLib.ClientOptions) Client) *Notifier {
+	if clientFactory == nil {
+		clientFactory = defaultClientFactory
+	}
+
+	opts := mqttLib.NewClientOptions().
+		AddBroker(cfg.BrokerURL).
+		SetClientID(cfg.ClientID).
+		SetUsername(cfg.Username).
+		SetPassword(cfg.Password)
+
+	if cfg.InsecureSkipVerify {
+		tlsCfg := tls.Config{
+			InsecureSkipVerify: true,
+		}
+		opts.SetTLSConfig(&tlsCfg)
+	}
+
+	return &Notifier{
+		Base:     receivers.NewBase(meta),
+		log:      logger,
+		tmpl:     template,
+		settings: cfg,
+		client:   clientFactory(opts),
+	}
+}
+
+// mqttMessage defines the JSON object send to an MQTT broker.
+type mqttMessage struct {
+	*templates.ExtendedData
+
+	// The protocol version.
+	Version  string `json:"version"`
+	GroupKey string `json:"groupKey"`
+	Message  string `json:"message"`
+}
+
+func (n *Notifier) Notify(ctx context.Context, as ...*types.Alert) (bool, error) {
+	n.log.Debug("Sending an MQTT message")
+
+	msg, err := n.buildMessage(ctx, as...)
```

Comment:
although we validate Message while parsing config, it does not guarantee that it cannot be empty here. Should we check the result for emptiness? 
It looks like MQTT does not put any requirements on this, though

Reply:
Yes, we can send even an empty message to a topic, not sure that we should check the result here.